File: Data Structures & Algorithms/combination-target-sum/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        let ret = []
        const dfs = ((i, subset, sumVal) =>{
            if(sumVal === target){
                ret.push([...subset])
                return
            }
            if(sumVal > target || i == nums.length){
                return
            }
            subset.push(nums[i])
            dfs(i, subset, sumVal + nums[i])
            subset.pop()
            dfs(i + 1, subset, sumVal)
        })
        dfs(0, [], 0)
        return ret

    }
}

File: Data Structures & Algorithms/subsets/submission-3.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ret = []
        def dfs(i, subset):
            if(i == len(nums)):
                ret.append(subset.copy())
                return
            
            subset.append(nums[i])
            dfs(i + 1, subset)
            subset.pop()
            dfs(i + 1, subset)
        
        dfs(0, [])
        return ret